src/coreclr/jit/objectalloc.h

```diff
@@ -33,7 +34,9 @@ class ObjectAllocator final : public Phase
     // definitely-stack-pointing pointers. All definitely-stack-pointing pointers are in both sets.
     BitVec              m_PossiblyStackPointingPointers;
     BitVec              m_DefinitelyStackPointingPointers;
+    BitVec              m_PointersHasLocalStore;
```

Comment:
Why does the analysis need to change for arrays? That seems odd.

Reply:
For example, the below code:
```cs
var a = new int[42];
a = new int[43];
```
will be imported to
```
***** BB01 [0000]
STMT00000 ( 0x000[E-] ... 0x007 )
               [000003] DACXG------                         *  STORE_LCL_VAR ref    V00 loc0
               [000002] --CXG------                         \--*  CALL help ref    CORINFO_HELP_NEWARR_1_VC
               [000001] H---------- arg0                       +--*  CNS_INT(h) long   0x7ffb1727d7b0 class int[]
               [000000] ----------- arg1                       \--*  CNS_INT   long   42

***** BB01 [0000]
STMT00001 ( 0x008[E-] ... 0x00F )
               [000007] DACXG------                         *  STORE_LCL_VAR ref    V00 loc0
               [000006] --CXG------                         \--*  CALL help ref    CORINFO_HELP_NEWARR_1_VC
               [000005] H---------- arg0                       +--*  CNS_INT(h) long   0x7ffb1727d7b0 class int[]
               [000004] ----------- arg1                       \--*  CNS_INT   long   43
```
where `V00` gets reused.

This can lead problems like:
- stack allocated arrays doesn't get proper zeroing before re-assignment
- setting it to `null` would lead to an assert in `RewriteUses` at https://github.com/dotnet/runtime/blob/a86987ccab917433d065fe5dc8870fc261f79d14/src/coreclr/jit/objectalloc.cpp#L907

Reply:
Yes, `V00` gets reused, but `V00` is just a pointer to the object. It is fine for it to be pointing to two different stack allocated objects at different times. The same case with a normal object should work just fine.

Reply:
Is this because you are trying to track the array length based on the pointer local (`V00`) here? I think that's the problem -- we should not try to do that. Instead, the IR of `INDEX_ADDR` should be transformed into something that preserves its semantics, yet exposes the potential array access to local morph. We do the same thing (in spirit) for `CORINFO_HELP_UNBOX`, which gets transformed into a `COMMA` that calls `CORINFO_UNBOX_HELPER_TYPETEST` on the LHS.

Reply:
> Is this because you are trying to track the array length based on the pointer local (V00) here?

No, I'm tracking the array length based on new stack allocated local.

Reply:
The issue here is that we were using `m_HeapLocalToStackLocalMap` to track the relationship between the original local and the new stack allocated local. When it sees a local which has been stack allocated (ie. a local has the relationship recorded in `m_HeapLocalToStackLocalMap`), it assumes the local must be a use and tries to update it to `LCL_VAR_ADDR`. But for arrays, it can be yet another `STORE_LCL_VAR` instead of a use.

Reply:
So essentially object stack allocation is relying on an IR invariant that all `ALLOCOBJ` nodes must be stored to different locals when we get to object allocation? And also that all uses have the store as its reaching def, and come in the right order in `Compiler::Blocks()`. This seems like a dangerous assumption to make in object stack allocation.

I do not think object stack allocation has a need for this anymore since local morph has proper flow sensitive propagation of these `LCL_ADDR`s now. Seems like something we should fix; the current approach is scary to me.

cc @AndyAyersMS, what do you think?

Reply:
The comment talks about exposing `ALLOCOBJ` as top level statements to be able to find them. That's ok with me; it is an optimization on throughput (though I question it being worth it, compared to just indicating its location with `GTF_CALL` or similar). If we didn't end up finding an `ALLOCOBJ` we just wouldn't stack allocate it.

That's not the only constraint the current approach has. It would also be illegal to optimize `newobj; stloc` into a top-level store directly to that local, or introduce any sort of transformation that does that between import and object stack allocation. If this happened object stack allocation could end up improperly replacing some uses of that local by a `LCL_ADDR`. That's a much more subtle invariant that the comment does not talk about.

Reply:
I guess I'm not that worried right now, though we should add this to the next steps list (which I need to resurrect).

The phase already asserts that all `ALLOCOBJ` are top level. We could extend that to insist that the store destination is at least marked as single-def temp (which is how the importer arranges things). That would catch some of the substitution issues, though it's not clear how far outside the importer one can trust single-def.

If we want to generalize the analysis or allow different IR shapes we will need to redesign things somewhat.

Reply:
> I think similar things need to be done for `newarr` as well.

Yes, for now you need to follow this pattern.